### PR TITLE
Convert spec command for virtctl

### PR DESCRIPTION
**Pull request by rmohr**
~Three things missing:~

 * ~The libvirt struct still depends on libvirt-go stuff~
 * ~Tests~
 * ~Documentation~

Example usage:

```bash
Convert between Kubevirt and Libvirt VM representations:

Examples:
# Convert Domain xml to yaml via stdin
virsh dumpxml testvm | virtctl convert-spec -f - -o yaml

# Convert yaml into json from a http resource
virtctl convert-spec -f http://127.0.0.1:4012/vm.yaml -o json

# Convert VM specification from a yaml file into a Libvirt Domain XML
virtctl convert-spec -f vm.yaml -o xml > dom.xml

Options:
  -f, --filename string         Filename, directory, or URL to files identifying the resource to get from a server.
  -o, --output string           Output format. One of: json|yaml|xml. (default "json")
      --output-version string   Output the formatted object with the given group version (default "kubevirt.io/v1alpha1")
```

The input format is automatically detected.

### CODE REVIEW

**Comment by rmohr** [comment]
/cc @fabiand @stu-gott @admiyo See above that it is not yet completely done, but the functionality is there and works.

Could you provide some feedback if the commandline arguments and the examples make sense in you eyes?

File: pkg/virtctl/convert/guess.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	}
+}
+
+func guessXMLStream(r io.Reader, size int) (io.Reader, []byte, bool) {
+	buffer := bufio.NewReaderSize(r, size)
+	b, _ := buffer.Peek(size)
+	return buffer, b, hasXMLPrefix(b)
+}
+
+var xmlPrefix = []byte("<")
```
Comment (stu-gott) [inline pkg/virtctl/convert/guess.go]: Is this truly sufficient?

File: pkg/virtctl/convert/guess.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"unicode"
+)
+
+func GuessStreamType(source io.Reader, size int) (io.Reader, Type) {
+	if r, _, isXML := guessXMLStream(source, size); isXML {
+		return r, XML
+	} else if r, _, isJSON := yaml.GuessJSONStream(r, size); isJSON {
+		return r, JSON
+	} else {
+		return r, YAML
```
Comment (stu-gott) [inline pkg/virtctl/convert/guess.go]: Is it possible to explicitly test for yaml rather than assuming at this point?

File: pkg/virtctl/convert/guess.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"k8s.io/client-go/pkg/util/yaml"
+	"unicode"
+)
+
+func GuessStreamType(source io.Reader, size int) (io.Reader, Type) {
```
Comment (stu-gott) [inline pkg/virtctl/convert/guess.go]: This function signature should probably also include an error code. For example there might be parse errors, or the stream doesn't match anything we know.

File: pkg/virtctl/convert/convert.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		return os.Stdin, nil
+	} else if strings.HasPrefix(sourceName, "http") {
+		resp, err := http.Get(sourceName)
+		if err != nil {
+			return nil, err
+		}
+		return resp.Body, nil
+
+	} else {
+		return os.Open(sourceName)
```
Comment (stu-gott) [inline pkg/virtctl/convert/convert.go]: Open returns (*File, error), is *File also a io.ReadCloser?

File: pkg/virtctl/convert/guess.go (hunk 3 above)
Comment (rmohr) [inline pkg/virtctl/convert/guess.go]: My assumption was, that the conversion later on will fail, if we have an unsupported function. I will think about it a little bit more.

File: pkg/virtctl/convert/guess.go (hunk 2 above)
Comment (rmohr) [inline pkg/virtctl/convert/guess.go]: Not really. For JSON, YAML and XML, we can just guess in every case. For JSON "{" (definitely not xml or yaml", for XML "<" (definitely not yaml or json).

If both don't match, I take the only possible left solution which will then later on fail when it is parsed.

File: pkg/virtctl/convert/convert.go (hunk 4 above)
Comment (rmohr) [inline pkg/virtctl/convert/convert.go]: Yes.

File: pkg/virtctl/convert/guess.go (hunk 1 above)
Comment (rmohr) [inline pkg/virtctl/convert/guess.go]: I think it is good enough. Everything more than that would involve parsing the whole xml (basically a validation then, which will be done by the decoder later on)

**Comment by michalskrivanek** [comment]
@rmohr why don't you want to depend on libvirt-go?

**Comment by rmohr** [comment]
> @rmohr why don't you want to depend on libvirt-go?

@michalskrivanek In the backend we do. virtctl is just a frontend commandline tool. If we would depend anywhere there directly on libvirt-go code, golang would link against the libvirt C library. We don't need anything libvirt specific in the frontend call.

**Comment by michalskrivanek** [comment]
ah, sure, thanks for clarification

**Comment by rmohr** [comment]
Ready for review

File: cmd/virtctl/virtctl.go [hunk 5]
```diff
@@ -53,10 +55,11 @@ func Parse(flags *flag.FlagSet) (*flag.FlagSet, error) {
 
 func Usage() {
 	fmt.Fprintln(os.Stderr,
-		`virtctl controll VM related operations on your kubernetes cluster.
+		`virtctl controls VM related operations on your kubernetes cluster.
 
 Basic Commands:
   console        Connect to a serial console on a VM
+  convert-spec   Convert between Libvirt and Kubevirt specifications
```
Comment (stu-gott) [inline cmd/virtctl/virtctl.go]: KubeVirt, right?

File: pkg/virtctl/convert/convert.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+func (c *Convert) FlagSet() *flag.FlagSet {
+	cf := flag.NewFlagSet("convert", flag.ExitOnError)
+	cf.StringP("filename", "f", "", "Filename, directory, or URL to files identifying the resource to get from a server.")
+	cf.StringP("output", "o", "json", "Output format. One of: json|yaml|xml.")
+	cf.String("output-version", "kubevirt.io/v1alpha1", "Output the formatted object with the given group version")
+	return cf
+}
+
+func (c *Convert) Usage() string {
+	usage := "Convert between Kubevirt and Libvirt VM representations:\n\n"
```
Comment (stu-gott) [inline pkg/virtctl/convert/convert.go]: KubeVirt

File: pkg/virtctl/convert/convert.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+func (c *Convert) Run(flags *flag.FlagSet) int {
+	sourceName, _ := flags.GetString("filename")
+
+	outputFlag, _ := flags.GetString("output")
+	outputFormat := Type(outputFlag)
+
+	if sourceName == "" {
+		log.Println("No source specified")
+		return 1
```
Comment (stu-gott) [inline pkg/virtctl/convert/convert.go]: It might be more clear if we create enums at the top of the module along the lines of "RuntimeSuccess" and "RuntimeError"

File: pkg/virtctl/convert/guess_test.go [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+	"bytes"
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	"io/ioutil"
+)
+
+var _ = Describe("Guess", func() {
+	table.DescribeTable("Should detect stram type", func(data string, detected Type) {
```
Comment (stu-gott) [inline pkg/virtctl/convert/guess_test.go]: "stream type"

File: pkg/virtctl/convert/convert.go (hunk 7 above)
Comment (rmohr) [inline pkg/virtctl/convert/convert.go]: Let's return an error or nil instead, how about that?

Comment (rmohr) [inline pkg/virtctl/convert/convert.go]: I would do that in a separate PR, since it is an interface change for all virtcl commands ...

Comment (rmohr) [inline pkg/virtctl/convert/convert.go]: Anyway, your suggestion is good. Done.

**Comment by stu-gott** [comment]
retest this please
